recipe/meta.yaml

```
@@ -24,17 +24,24 @@ outputs:
     requirements:
       host:
         # Needed to symlink libcuda into sysroot libs.
-        - {{ compiler("c") }}
+        # FIXME: nvcc 9.* and 10.* don't work with GCC 9.
+        #        Adding custom GCC versions via conda_build_config.yaml fails
+        #        due to https://github.com/conda/conda-build/issues/3974
+        #        => use hardcoded compiler package names to workaround that.
+        - {{ compiler("c") }}    # [not (linux64 and (cuda_compiler_version or "").startswith(("9.", "10.")))]
+        - gcc_linux-64 7         # [linux64 and (cuda_compiler_version or "").startswith(("9.", "10."))]
```

Does this have a `run_export`? If so can we do `<= 7` so users can use an older compiler if desired still?

Similarly, it would be good to handle the CUDA 11.0, and 11.1 requirements: https://github.com/conda-forge/nvcc-feedstock/issues/51

Something like:
```suggestion
        - gcc_linux_64 <=7       # [linux64 and (cuda_compiler_version or "").startswith(("9.", "10."))]
        - gcc_linux_64 <=9       # [linux64 and (cuda_compiler_version or "").startswith("11.0")]
        - gcc_linux_64 <=10     # [linux64 and (cuda_compiler_version or "").startswith("11.1")]
```

---

This is just the compiler version used during the build/test. We only have `run_exports` for the runtimes (`libgcc-ng`/`libstdcxx-ng`) but not for the compilers themselves. You should be able to build downstream packages with older compilers either way.
I've used the exact versions because that's how we build packages on conda-forge generally, with a fixed compiler version (`7` or `9` recently). I've also not special cased `nvcc=10.1|10.2` to have a consistent version with other packages (we hadn't pinned GCC `8` before).

My goal was to do the least necessary for the build and rather wait until someone fixes the mentioned conda-build bug instead of investing time in workarounds.
That said, I think it's best if I split this change off into another PR and just use a temporary workaround in the CI files for the intended change of this PR.

---

I hadn't looked at the whole recipe until now...
I don't really understand the purpose of the `{{ compiler('c') }}` dependency in `host` here. It is not used in any way.
Also the other duplicate deps for sysroot/glibc I dont understand either.

I can't deduce what the intentions of the dependency declarations overall are. They need some cleanup, but without knowing the indented use, I can't help with that.

---

`nvcc` uses the host compiler during it's compilation. Depending on the CUDA version, there's compatibilities with the version of the host compiler used that we need to respect.

Additionally, as of CUDA 11, CentOS 6 is no longer supported so we needed a way to force only CentOS 7+.

---

> `nvcc` uses the host compiler during it's compilation. Depending on the CUDA version, there's compatibilities with the version of the host compiler used that we need to respect.

This recipe only creates wrapper/(de-)activation scripts => the compiler is not used. If it's meant to be declared for downstream packages, something has to be declared in `requirements/run`.

> Additionally, as of CUDA 11, CentOS 6 is no longer supported so we needed a way to force only CentOS 7+.

The duplication of that in `host`/`run`/`test/requires` is what I meant.